### DISCUSSION TITLE
Technology debt

### DISCUSSION THREAD

**Discussion by Tatrabbit** [SillyTavern/SillyTavern#2937] [General]
### Have you searched for similar requests?

Yes

### Is your feature request related to a problem? If so, please describe.

I've begun tinkering with the source code, and it's become clear to me that there are a few design problems that are holding back the speed at which code can be developed. The Generate() function inside of script is a primary example of this. It's a large, monolithic function that has lots of hacky workarounds depending on the main_api that's selected, for instance.

### Describe the solution you'd like

There's a very high amount of technology debt, and encapsulation has broken down almost entirely.

The LLM API's should be refactored to interface/wrap the actual underlying API. that would eliminate the need to check which kind of API is being used outside of it.

Author's Note is a kind of World Info, so it could share code.

UI is entangled with function throughout the project.

Making these changes would be a time-consuming, error-prone process that would make merging outstanding pull requests very difficult. However, it would greatly ease future development, enable more complicated features to be added more quickly, because regressions would happen far less frequently and require less knowledge of the full project's design.

### Describe alternatives you've considered

I've considered forking the repository as my own, but it would be much nicer if my improvements could be merged back and I'd like to coordinate that.

### Priority

Medium (Would be very useful)

### Are you willing to test this on staging/unstable branch if this is implemented?

Yes

**Comment by Cohee1207** [comment]
> I've considered forking the repository as my own, but it would be much nicer if my improvements could be merged back and I'd like to coordinate that.

You're free to. But it seems easier to start anew than to try and fix what's already there.

**Comment by Tatrabbit** [comment]
That reply sounds like you, yourself, are rather frustrated with the state of things. ST has an absolute freighter of features and it would be a shame to have Yet Another Tavern out there. I'll try to find something to chip away at slowly. Is there any major thing you'd prefer done sooner rather than later? I'll probably start with something very small but easily fixable like JQuery being rampantly invoked without caching.

Edit: I feel I misread. Sorry. I'll fork if you'd prefer that.

**Comment by Cohee1207** [comment]
I'm aware that the code base is far from perfect and I'm fine with accepting small fixes. However large refactorings (like a complete rewrite of Generate()) are nearly impossible at this point. That's why my point is that doing it right, but from scratch, is more feasible. Some of the technological flaws drag from the parent project.

If you wish to coordinate, please contact me via Discord (see README), that would be easier to do than here.

**Comment by Tatrabbit** [comment]
I would have to agree. Right now, I've begun work on a system-wide state manager. I want to make it interfere as little as possible with existing code. Even so, I'm a little bit nervous because I'm not the sociable sort, and although I do feel confident enough to write code I personally want, I'm either suffering from Imposter Syndrome or maybe not. Can't tell!

I took a short pause from the task since I have other coding projects that had a more immediate need to address, but what I think the next steps should be is to at least push my commits to my fork and have people look over them. Some reassurance that I'm just doing something along the correct line would help.

**Comment by Wolfsblvt** [comment]
Sure, having your own fork and branch and having a Draft PR can help keeping track of changes and discuss/comment on this, if that's what you want.

Still, I am with Cohee, actually coordinating and bouncing ideas and concepts off of each other via brainstorming is something you can do much better in a chat, than having GitHub issues with several dozen or even hundreds of comments. They aren't really designed to *chat*.

For many things that exist there might also be reasons more than "this is just old code/how it is", that might not be obvious to devs that aren't used to ST or know it in depth.
That's again where a lot of devs have profited to being able to quickly ask questions in the "code-contributions" channel on discord and can get help, replies and feedback there.
Not trying to say Discord is in any way required. Contributions and discussions via GitHub are totally welcome. Just saying it's much easier to chat, especially if you want to get deeper into the actual framework for ST.

**Comment by Tatrabbit** [comment]
Yeah, I'm sorry about this. I've not had much luck trying to find a way that'll truly solve the problem other than just move it into a new container. I think that a completely new software would be needed as Cohee originally said. Maybe given enough time, the strangler fig approach would work, but, at that point, I'd rather write my own. SillyTavern has given me so much inspiration and fun and I'd like to contribute meaningfully. Clearly being annoyed at a really multi-fingered problem is not the way to go.
